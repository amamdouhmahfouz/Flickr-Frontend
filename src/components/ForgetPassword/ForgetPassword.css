div.forgetpassword-page {

    background-size:cover;
    background-position:inherit;
    padding-top: 1px;
    padding-bottom: 10%;
    background-image: url(bg22.jpg);
    background-attachment:fixed;

}


img.forgotpass {
    height: auto;
    width: 70px;
    position: relative;
    
}

.searching {
    display: flex;
    align-items: center;
    padding-left: 20%;
}