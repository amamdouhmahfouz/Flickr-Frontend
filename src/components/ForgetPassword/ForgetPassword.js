import React, { useState, useEffect } from 'react'
import searchImg from './search1.png'
import axios from 'axios'
import { Redirect } from "react-router-dom";
import classes from './ForgetPassword.module.css'
import passImg from './icon-password.jpg'
import API from '../../fakeAPI';


/**
 * A from that takes email as input and searches for it in the database
 * If found the user should be sent verification email, to be able to go to the reset password form  
 * 
 * @author Esraa Hamed
 * @async
 * @example <ForgetPassword />
 * @returns {element} The Forget Password form contents
 *
 */
const ForgetPassword = () => {

    const apiURL = "http://localhost:3000/users" ;   //json server
    const [passwordURL, setPasswordURL] = useState();
    const [redirect, setRedirect] = useState(null);
    const [userError, setUserError] = useState('');

    const [email, setEmail] = useState('');
    const [emailError, setemailError] = useState();
    const [isSubmitting, setisSubmitting] = useState(false);
    
    /**
     * Handles what happens when form is submitted
     * 
     * @param {object} e - the JavaScript event object
     */
    const handleSubmit = (e) => {
        e.preventDefault();
        checkUserInput();
        validateEmail();
        setisSubmitting(true);
    }
    
    //---------------------------------------- HANDLING INPUTS ---------------------------------------//

    const handleEmailInput = (e) => {
        setEmail(e.target.value);  
    }
    
    // --------------------------------------- VALIDATIONS ------------------------------------------ //

    const validateEmail = () => {
        if(!email){
            setemailError('Email is required'); setUserError('');      
        }
        else {setemailError('')}
    }

    // ---------------------------------------- json server -------------------------------------------//
    /**
     * Checks if the email inputted by the user is in the database
     * If found, the user will be directed another form showing a message to check his/her email for validation
      * Otherwise, the user remains in forget password form page and is shown an error message : 'Invalid email' 
     */
    const checkUserInput = () => {
      API.get('users?email=' + email) 
      .then(response => {
         console.log(response);
         if(response.data.length > 0 ) {
             setemailError('');
             setPasswordURL('localhost3001/forgotpassword-confrim?email=' + email )
             setRedirect("/post-forgotpassword");
         } else if (response.data.length ===0 && email) {
             setUserError('Invalid email');
         }
      })
    }

    // ------------------------------------------ RETURN -------------------------------------------------- //

     if(redirect) {
        return <Redirect to={redirect} />
      }

    return (
        <div className={classes.div__forgetpassword_page}  data-testid="forgetpassword" >           
        <form className={classes.form__forgetpassword} onSubmit={handleSubmit} data-testid="form">
              
               <img className={classes.img__forgotpass} src={passImg} /> 
              <h4 className={classes.h4__center}> Forgot your password ? </h4>
              <h6 className={classes.h6__center}> Please enter your email to search for your account </h6>

              <div className={classes.div__input}>
                  <input type="email" placeholder="Email address" className={classes.div__inputfield} id="email"
                         onChange={handleEmailInput} value={email} data-testid="input"/>
                         <p className={classes.p__error}>{emailError}</p>
              </div>

              <div className={classes.div__usererror}>
              <p className={classes.p__error2}>{userError}</p>
              </div>
                   
              <div className={classes.searching}>
                 <button className={classes.div_forgetpasswordbutton}  id="forgotpassword-button" data-testid="button"> Search </button>
              </div>
                 
                 <br />
                 <br />
               </form>
        </div>
    );
}

export default ForgetPassword