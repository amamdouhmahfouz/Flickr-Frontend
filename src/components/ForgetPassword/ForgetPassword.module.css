.div__forgetpassword_page {

    background-size:cover;
    background-position:inherit;
    padding-top: 1px;
    padding-bottom: 10%;
    background-image: url(bg22.jpg);
    background-attachment:fixed;

}
.div__resetpassword_page {

    background-size:cover;
    background-position:inherit;
    padding-top: 1px;
    padding-bottom: 10%;
    background-image: url(bg22.jpg);
    background-attachment:fixed;
    padding-top: 7%;
}


.img__forgotpass {
    height: auto;
    width: 70px;
    position: relative;
    
}



.h4__center
{
    color: black;
}
.btn__block {
    width: 60%;
}
.searching {
    display: flex;
    align-items: center;
    padding-left: 20%;
}