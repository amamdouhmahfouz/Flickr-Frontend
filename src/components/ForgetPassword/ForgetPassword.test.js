import React from 'react' ;
import ReactDOM from 'react-dom' ;
import ForgetPassword from '../ForgetPassword/ForgetPassword'

import {render, cleanup, fireEvent} from '@testing-library/react'
import '@testing-library/jest-dom'
import renderer from 'react-test-renderer'

afterEach(cleanup);

it("renders without crashing", ()=> {
    const div = document.createElement("div");
    ReactDOM.render(<ForgetPassword></ForgetPassword>, div)
})

it("matches snapshot", () => {
    const tree = renderer.create(<ForgetPassword></ForgetPassword>).toJSON();
    expect(tree).toMatchSnapshot();
})

describe("Input value", () => {
    it("updates on change", ()=> {
        const {queryByTestId} = render(<ForgetPassword />) 
        const emailInput = queryByTestId("input");
        fireEvent.change(emailInput, {target: {value:"test"}})
        expect(emailInput.value).toBe("test");
    })
})

test('Empty fields validation', () => {

    const {queryByTestId, getByText} = render(<ForgetPassword />);
    const emailInput = queryByTestId("input");
    const ForgetPasswordForm = queryByTestId("form");
    fireEvent.change(emailInput, {target: {value:""}})
    fireEvent.submit(ForgetPasswordForm);
    expect(getByText('Email is required')).toBeInTheDocument();
})

test('Filled fields validation', () => {

    const {queryByTestId, queryByText} = render(<ForgetPassword />);
    const emailInput = queryByTestId("input");
    const ForgetPasswordForm = queryByTestId("form");
    fireEvent.change(emailInput, {target: {value:"test@test.test"}})
    fireEvent.submit(ForgetPasswordForm);
    expect(queryByText('Email is required')).not.toBeInTheDocument();
})


