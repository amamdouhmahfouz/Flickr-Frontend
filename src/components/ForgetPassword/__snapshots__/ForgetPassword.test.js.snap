// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<div
  className="div__forgetpassword_page"
  data-testid="forgetpassword"
>
  <form
    className="form__forgetpassword"
    data-testid="form"
    onSubmit={[Function]}
  >
    <img
      className="img__forgotpass"
      src="icon-password.jpg"
    />
    <h4
      className="h4__center"
    >
       Forgot your password ? 
    </h4>
    <h6
      className="h6__center"
    >
       Please enter your email to search for your account 
    </h6>
    <div
      className="div__input"
    >
      <input
        className="div__inputfield"
        data-testid="input"
        id="email"
        onChange={[Function]}
        placeholder="Email address"
        type="email"
        value=""
      />
      <p
        className="p__error"
      />
    </div>
    <div
      className="div__usererror"
    >
      <p
        className="p__error2"
      >
        
      </p>
    </div>
    <div
      className="searching"
    >
      <button
        className="div_forgetpasswordbutton"
        data-testid="button"
        id="forgotpassword-button"
      >
         Search 
      </button>
    </div>
    <br />
    <br />
  </form>
</div>
`;
